### ISSUE TITLE
performance issue ~30mins loading WMT dataset, transformer

### ISSUE THREAD

**Issue by vsuthichai** [NVIDIA/OpenSeq2Seq#202]
I'm attempting to train a transformer-big model with 16gpus across 2 nodes (8 gpus each) on AWS with Horovod  and experiencing a performance issue when the run.py script starts up.  With the tensorflow timeline, I've narrowed it down to what I believe is the `IteratorGetNext` op.  I suspect it's spending a lot of time trying to load and preprocess the dataset.  This is on an EBS backed instance.

![screen shot 2018-08-01 at 6 12 54 pm](https://user-images.githubusercontent.com/2184688/43556826-9eb927a4-95b6-11e8-8afc-85fcda1072d8.png)

**Comment by vsuthichai** [comment]
An update.. this happens even when i use 2 gpus on a single node.

**Comment by okuchaiev** [comment]
We've recently updated the code to address this issue:

1) Please get latest from master
2) Re-run https://github.com/NVIDIA/OpenSeq2Seq/blob/master/get_wmt16_en_dt.sh  script or only part of it which does full data shuffle
3) Make sure that in your training config you are setting a shuffle buffer: https://github.com/NVIDIA/OpenSeq2Seq/blob/master/example_configs/text2text/en-de/transformer-big.py#L103 

What will be happening is: (1) the data will undergo full shuffle before any training is started.
(2) during training, the data shuffle buffer will shuffle data in that rolling buffer, not full dataset

**Comment by vsuthichai** [comment]
@okuchaiev So the fix worked, though I'm curious to understand in depth as to why this worked.  According to this post, https://stackoverflow.com/questions/46444018/meaning-of-buffer-size-in-dataset-map-dataset-prefetch-and-dataset-shuffle#, the buffer specifies the amount of data to be copied from the dataset into a buffer, where samples are selected at random from that buffer.  If the buffer size is the size of the dataset, then it sounds like essentially samples are selected at random from the entire dataset instead of the buffer.  How does the slow down surface here?

**Comment by okuchaiev** [comment]
well, before the fix it would shuffle ALL the data after every epoch. This actually isn't too much data - only around 4.5 sentence pairs and could even fit in RAM, but somehow tf.data does it very slowly.

On a side note - you do need to shuffle your training data.
